Add support for X.509 in demo server

scop opened the pull request:
Tested successfully with `output/bin/demo` and the generated certs from `output/certs` with current https://github.com/AVSystem/Anjay master. See also https://github.com/AVSystem/Anjay/pull/7

jvermillard commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                        if (keyStore.isCertificateEntry(alias)) {
+                            trustedCertificates.add(keyStore.getCertificate(alias));
+                        } else if (keyStore.isKeyEntry(alias) && alias.equals(keyStoreAlias)) {
+                            ArrayList<X509Certificate> x509CertificateChain = new ArrayList<>();
```
List<...>

jvermillard commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                KeyStore keyStore = KeyStore.getInstance(keyStoreType);
+                try (FileInputStream fis = new FileInputStream(keyStorePath)) {
+                    keyStore.load(fis, keyStorePass == null ? null : keyStorePass.toCharArray());
+                    ArrayList<Certificate> trustedCertificates = new ArrayList<>();
```
List<...>

jvermillard commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                                    if (certificate instanceof X509Certificate) {
+                                        x509CertificateChain.add((X509Certificate) certificate);
+                                    } else {
+                                        LOG.warn("Ignoring non-X.509 certificate in alias chain: {}", certificate);
```
this is probably an error if you have a non-X509 cert in the cert-chain, so an exception or any fatal error is better

jvermillard commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                                            key == null ? null : key.getClass().getName());
+                                    System.exit(-1);
+                                }
+                            } else {
```
move the error case before and put the nominal case outside the if

jvermillard commented:
Thanks! Would be nice to have the command line you used for generating the key-store, so we can put it in the instructions.

sbernard31 commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                                if (key instanceof PrivateKey) {
+                                    builder.setPrivateKey((PrivateKey) key);
+                                    publicKey = keyStore.getCertificate(alias).getPublicKey();
+                                    builder.setPublicKey(publicKey);
```
You don't need set public key as explain in LeshanServer javadoc. We should add javadoc to builder too.

scop commented:
I think this addresses all the comments. Did not have time to re-test yet though.

I don't have the exact commands I used to create the keystore. What I basically did was took serverKeyStore.jks from Leshan integration tests, and added root.crt generated during Anjay's build to it using Portecle (portecle.sf.net, shameless plug).

scop commented:
(Tested briefly now, still works for me.)

sbernard31 commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -194,35 +225,88 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                            } else {
+                                LOG.error("Keystore alias must have a PrivateKey entry, was {}",
+                                        key == null ? null : key.getClass().getName());
+                                System.exit(-1);
```
move the error case before and put the nominal case outside the if

sbernard31 commented on leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:
```diff
@@ -194,35 +225,88 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
+                                    x509CertificateChain.add((X509Certificate) certificate);
+                                } else {
+                                    LOG.error("Non-X.509 certificate in alias chain is not supported: {}", certificate);
+                                    System.exit(-1);
```
move the error case before and put the nominal case outside the if

scop commented:
Remaining comments addressed now.
